File: facesdk_java_tests/src/test/java/GroupTests.java
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import com.regula.facesdk.webclient.ApiException;
import com.regula.facesdk.webclient.gen.model.*;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;

class GroupTests {

    private UUID groupId;
    private Group groupResponse;
    private final String testGroupName = "test";
    private final String nameA = "Person A";
    Map<String, Object> baseMetadata = new HashMap<>() {{
        put("description", "This is a test group");
    }};

    @BeforeEach
    void setUp() {
        GroupToCreate groupToCreate = new GroupToCreate();
        groupToCreate.setName(testGroupName);
        groupToCreate.setMetadata(baseMetadata);
        groupResponse = PathsConfig.faceSdk.groupApi.createGroup(groupToCreate);
        Assertions.assertNotNull(groupResponse.getId(), "No id field in the returned group");
        groupId = groupResponse.getId();
    }


    @AfterEach
    void tearDown() {
        if (groupId != null) {
            try {
                PathsConfig.faceSdk.groupApi.deleteGroup(groupId);
            } catch (ApiException e) {
                System.out.println("Error during group deletion: " + e.getMessage());
                // Handle other necessary cleanup, if any
            }
        }
    }


    @Test
    void shouldCreateAndDeleteGroup() {
        try {
            Assertions.assertNotNull(groupResponse, "No response received");
            Assertions.assertEquals(testGroupName, groupResponse.getName(),
                    "Expected group name to be " + testGroupName +
                            ", but got " + groupResponse.getName() + " instead.");
            Assertions.assertEquals(baseMetadata.get("description"),
                    groupResponse.getMetadata().get("description"),
                    "Expected metadata description to be " + baseMetadata +
                            ", but got " + groupResponse.getMetadata() + " instead.");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
        }
    }

    @Test
    void shouldGetAllGroups() {
        List<UUID> createdGroupIds = new ArrayList<>();
        createdGroupIds.add(groupId);

        try {
            // Creating additional groups
            for (int i = 1; i < 3; i++) {
                GroupToCreate additionalGroupToCreate = new GroupToCreate();
                additionalGroupToCreate.setName(testGroupName);
                additionalGroupToCreate.setMetadata(baseMetadata);
                Group additionalGroup = PathsConfig.faceSdk.groupApi.createGroup(additionalGroupToCreate);
                UUID additionalGroupId = additionalGroup.getId();
                Assertions.assertNotNull(additionalGroupId, "No id field in the returned group");
                createdGroupIds.add(additionalGroupId);
            }

            // Getting all groups for a specific page
            GroupPage response = PathsConfig.faceSdk.groupApi.getAllGroups(2, 2);

            // Assertions
            Assertions.assertNotNull(response, "No response received");
            Assertions.assertEquals(2, response.getItems().size(), "Expected 2 items on page, but got different items count");
            Assertions.assertEquals(2, response.getPage(), "Expected page to be 2, but got different page number");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        } finally {
            // Clean up - delete all created groups
            for (UUID id : createdGroupIds) {
                try {
                    PathsConfig.faceSdk.groupApi.deleteGroup(id);
                } catch (ApiException e) {
                    System.err.println("Failed to delete group with ID " + id + ": " + e.getResponseBody());
                }
            }
        }
    }


    @Test
    void shouldGetAllPersonsByGroupId() {
        try {
            // Create a person with metadata
            PersonFields personFields = new PersonFields();
            personFields.setName(nameA);
            personFields.setGroups(List.of(groupId));
            personFields.setMetadata(baseMetadata);

            // Create a person
            Person createdPerson = PathsConfig.faceSdk.personApi.createPerson(personFields);
            UUID personId = createdPerson.getId();

            // Get all persons by groupId
            PersonsPage personsResponse = PathsConfig.faceSdk.groupApi.getAllPersonsByGroupId(1, 1, groupId);
            Assertions.assertNotNull(personsResponse, "No response received");

            // Check if the created person is in the group
            Assertions.assertTrue(
                    Objects.requireNonNull(personsResponse.getItems()).stream()
                            .anyMatch(person -> Objects.equals(person.getId(), personId)),
                    "Created person not found in the group"
            );
            Person retrievedPerson = personsResponse.getItems().stream()
                    .filter(person -> Objects.equals(person.getId(), personId))
                    .findFirst()
                    .orElse(null);

            Assertions.assertNotNull(retrievedPerson, "Retrieved person is null");
            Assertions.assertEquals(nameA, retrievedPerson.getName(),
                    "Name of Person is incorrect");
            Assertions.assertEquals(
                    baseMetadata.get("description"),
                    retrievedPerson.getMetadata().get("description"),
                    "Metadata of the group is incorrect"
            );
            Assertions.assertTrue(
                    retrievedPerson.getGroups().contains(groupId),
                    "group_id is incorrect"
            );

        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }

    @Test
    void shouldGetAGroup() {
        try {
            // Get all groups and take the first group_id that appears
            GroupPage groupsResponse = PathsConfig.faceSdk.groupApi.getAllGroups(1, 1);
            Assertions.assertTrue(Objects.requireNonNull(groupsResponse.getItems()).size() > 0, "No groups found");
            UUID firstGroupId = groupsResponse.getItems().get(0).getId();

            // Using obtained id to get chosen group
            Group groupResponse = PathsConfig.faceSdk.groupApi.getGroup(firstGroupId);
            Assertions.assertNotNull(groupResponse, "No response received");
            Assertions.assertEquals(groupResponse.getId(), firstGroupId,
                    "group_id is incorrect");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
        }
    }

    @Test
    void shouldUpdateAGroup() {
        try {
            // Updating group with a new name and metadata
            String newName = "updated_test_group";
            Map<String, Object> updatedMetadata = new HashMap<>() {{
                put("description", "updated meta");
            }};
            GroupToCreate updatedGroupToCreate = new GroupToCreate();
            updatedGroupToCreate.setName(newName);
            updatedGroupToCreate.setMetadata(updatedMetadata);
            PathsConfig.faceSdk.groupApi.updateGroup(groupId, updatedGroupToCreate);
            // Fetching the updated group and verifying the update
            Group updatedGroupResponse = PathsConfig.faceSdk.groupApi.getGroup(groupId);
            Assertions.assertEquals(newName, updatedGroupResponse.getName(),
                    "Group name wasn't updated");
            Assertions.assertEquals(updatedMetadata.get("description"),
                    updatedGroupResponse.getMetadata().get("description"),
                    "Group metadata wasn't updated");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
        }
    }

    @Test
    void shouldUpdatePersonsInGroup() {
        try {
            // Creating some persons and adding them to the group
            PersonFields personFieldsA = new PersonFields();
            personFieldsA.setName(nameA);
            personFieldsA.setGroups(Collections.singletonList(groupId));
            personFieldsA.setMetadata(baseMetadata);
            PathsConfig.faceSdk.personApi.createPerson(personFieldsA);
            // Getting the group's persons
            PersonsPage initialPersonsResponse = PathsConfig.faceSdk.groupApi.getAllPersonsByGroupId(1, 10, groupId);
            List<UUID> initialPersonIds = Objects.requireNonNull(initialPersonsResponse.getItems()).stream()
                    .map(Person::getId)
                    .collect(Collectors.toList());

            // Removing them by updating group
            UpdateGroup updateGroup = new UpdateGroup();
            updateGroup.setRemoveItems(initialPersonIds);
            PathsConfig.faceSdk.groupApi.updatePersonsInGroup(groupId, updateGroup);

            // Creating new persons and adding them to the group
            List<String> newPersonNames = new ArrayList<>(List.of("New Person A", "New Person B"));
            List<UUID> newPersonIds = new ArrayList<>();
            for (String name : newPersonNames) {
                PersonFields newPersonFields = new PersonFields();
                newPersonFields.setName(name);
                newPersonFields.setGroups(Collections.singletonList(groupId));
                newPersonFields.setMetadata(baseMetadata);
                UUID newPersonId = PathsConfig.faceSdk.personApi.createPerson(newPersonFields).getId();
                newPersonIds.add(newPersonId);
            }

            // Testing that they're here
            PersonsPage updatedPersonsResponse = PathsConfig.faceSdk.groupApi.getAllPersonsByGroupId(1, 10, groupId);
            List<UUID> updatedPersonIds = new ArrayList<>();
            for (Person person : Objects.requireNonNull(updatedPersonsResponse.getItems())) {
                UUID id = person.getId();
                updatedPersonIds.add(id);
            }
            List<String> updatedPersonNames = new ArrayList<>();
            for (Person person : Objects.requireNonNull(updatedPersonsResponse.getItems())) {
                String name = person.getName();
                updatedPersonNames.add(name);
            }

            // Sort the arrays before comparing
            Collections.sort(updatedPersonIds);
            Collections.sort(newPersonIds);
            Collections.sort(updatedPersonNames);
            Collections.sort(newPersonNames);

            Assertions.assertEquals(newPersonIds, updatedPersonIds, "Updated person ids don't match");
            Assertions.assertEquals(newPersonNames, updatedPersonNames, "Updated person names don't match");
            Assertions.assertEquals(2, updatedPersonIds.size(), "Expected 2 persons in updated list but got a different number");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getMessage());
        }
    }

    @Test
    void ShouldCreateGroupWithoutMetadata() {
        GroupToCreate groupToCreate = new GroupToCreate();
        groupToCreate.setName(testGroupName);
        groupResponse = PathsConfig.faceSdk.groupApi.createGroup(groupToCreate);
        UUID newGroupId = groupResponse.getId();
        PathsConfig.faceSdk.groupApi.deleteGroup(newGroupId);
    }
}

File: facesdk_cs_tests/facesdk_cs_tests/GroupTests.cs
namespace facesdk_cs_tests;

[TestFixture]
public class GroupTests
{
    private Guid? _groupId;
    private Group? _groupResponse;
    private const string TestGroupName = "test";
    private const string NameA = "Person A";

    private readonly Dictionary<string, object> _baseMetadata = new Dictionary<string, object>
    {
        { "description", "This is a test group" }
    };

    [SetUp]
    public void SetUp()
    {
        var groupToCreate = new GroupToCreate
        {
            Name = TestGroupName,
            Metadata = _baseMetadata
        };
        _groupResponse = PathsConfig.FaceSdk.GroupApi.CreateGroup(groupToCreate);
        _groupId = _groupResponse.Id;
    }

    [TearDown]
    public void TearDown()
    {
        if (!_groupId.HasValue) return;
        try
        {
            PathsConfig.FaceSdk.GroupApi.DeleteGroup(_groupId.Value);
        }
        catch (ApiException e)
        {
            Console.WriteLine("Error during group deletion: " + e.Message);
        }
    }

    [Test]
    public void ShouldCreateAndDeleteGroup()
    {
        Assert.That(_groupResponse, Is.Not.Null, "No response received");
        if (_groupResponse == null) return;
        Assert.Multiple(() =>
        {
            Assert.That(_groupResponse.Name, Is.EqualTo(TestGroupName),
                $"Expected group name to be {TestGroupName}, but got {_groupResponse.Name} instead.");
            Assert.That(_groupResponse.Metadata.TryGetValue("description", out var value),
                Is.EqualTo(_baseMetadata.TryGetValue("description", out var value1)),
                $"Expected group name to be {value1}, but got {value} instead.");
        });
    }

    [Test]
    public void ShouldGetAllGroups()
    {
        Debug.Assert(_groupId != null, nameof(_groupId) + " != null");
        var createdGroupIds = new List<Guid> { _groupId.Value };

        try
        {
            // Creating additional groups
            for (var i = 1; i < 3; i++)
            {
                var additionalGroupToCreate = new GroupToCreate
                {
                    Name = TestGroupName,
                    Metadata = _baseMetadata
                };
                var additionalGroup = PathsConfig.FaceSdk.GroupApi.CreateGroup(additionalGroupToCreate);
                var additionalGroupId = additionalGroup.Id;
                createdGroupIds.Add(additionalGroupId);
            }

            // Getting all groups for a specific page
            var response = PathsConfig.FaceSdk.GroupApi.GetAllGroups(2, 2);

            // Assertions
            Assert.That(response, Is.Not.Null, "No response received");
            Assert.That(response.Items, Is.Not.Null, "No items in the response");
            Assert.Multiple(() =>
            {
                Assert.That(response.Items, Has.Count.EqualTo(2),
                    "Expected 2 items on page, but got different items count");
                Assert.That(response.Page, Is.EqualTo(2), "Expected page to be 2, but got different page number");
            });
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
        finally
        {
            // Clean up - delete all created groups
            foreach (var id in createdGroupIds)
            {
                try
                {
                    PathsConfig.FaceSdk.GroupApi.DeleteGroup(id);
                }
                catch (ApiException e)
                {
                    Console.Error.WriteLine($"Failed to delete group with ID {id}: {e.ErrorCode + e.Message}");
                }
            }

            _groupId = null;
        }
    }

    [Test]
    public void ShouldGetAllPersonsByGroupId()
    {
        try
        {
            // Create a person with metadata
            var personFields = new PersonFields
            {
                Name = NameA,
                Groups = new List<Guid> { _groupId!.Value },
                Metadata = _baseMetadata
            };

            // Create a person
            var createdPerson = PathsConfig.FaceSdk.PersonApi.CreatePerson(personFields);
            var personId = createdPerson.Id;

            // Get all persons by groupId
            var personsResponse = PathsConfig.FaceSdk.GroupApi.GetAllPersonsByGroupId(1, 1, _groupId.Value);
            Assert.That(personsResponse, Is.Not.Null, "No response received");
            Assert.That(personsResponse.Items.Any(person => person.Id == personId), Is.True,
                "Created person not found in the group");
            Assert.Multiple(() =>
            {
                Assert.That(personsResponse.Items.Any(person => person.Name == NameA), Is.True,
                    "Name of Person is incorrect");
                Assert.That(personsResponse.Items.Any(person => person.Groups[0] == _groupId), Is.True,
                    "GroupId of Person is incorrect");
            });
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
    }

    [Test]
    public void ShouldGetAGroup()
    {
        try
        {
            // Get all groups and take the first group_id that appears
            var groupsResponse = PathsConfig.FaceSdk.GroupApi.GetAllGroups(1, 1);
            Assert.That(groupsResponse.Items?.Count, Is.GreaterThan(0), "No groups found");

            Debug.Assert(groupsResponse.Items != null, "groupsResponse.Items != null");
            var firstGroupId = groupsResponse.Items[0].Id;
            _groupResponse = PathsConfig.FaceSdk.GroupApi.GetGroup(firstGroupId);
            Assert.That(_groupResponse, Is.Not.Null, "No response received");
            Assert.That(_groupResponse.Id, Is.EqualTo(firstGroupId),
                $"Expected group name to be {firstGroupId}, but got {_groupResponse.Id} instead.");
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
        }
    }

    [Test]
    public void ShouldUpdateAGroup()
    {
        try
        {
            // Updating group with a new name and metadata
            const string newName = "updated_test_group";

            Dictionary<string, object> updatedMetadata = new Dictionary<string, object>
            {
                { "description", "This is a test group" }
            };

            var updatedGroupToCreate = new GroupToCreate
            {
                Name = newName,
                Metadata = updatedMetadata
            };

            Debug.Assert(_groupId != null, nameof(_groupId) + " != null");
            PathsConfig.FaceSdk.GroupApi.UpdateGroup(_groupId.Value, updatedGroupToCreate);

            // Fetching the updated group and verifying the update
            Debug.Assert(_groupId != null, nameof(_groupId) + " != null");
            var updatedGroupResponse = PathsConfig.FaceSdk.GroupApi.GetGroup(_groupId.Value);
            Assert.That(updatedGroupResponse.Name, Is.EqualTo(newName), "Group name not updated");
            Assert.That(updatedGroupResponse.Metadata.TryGetValue("description", out var value),
                Is.EqualTo(updatedMetadata.TryGetValue("description", out var value1)),
                $"Expected group name to be {value1}, but got {value} instead.");
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
        }
    }

    [Test]
    public void ShouldUpdatePersonsInGroup()
    {
        try
        {
            // Creating some persons and adding them to the group
            Debug.Assert(_groupId != null, nameof(_groupId) + " != null");
            var personFieldsA = new PersonFields
            {
                Name = NameA,
                Groups = new List<Guid> { _groupId.Value },
                Metadata = _baseMetadata
            };

            PathsConfig.FaceSdk.PersonApi.CreatePerson(personFieldsA);

            // Getting the group's persons
            var initialPersonsResponse = PathsConfig.FaceSdk.GroupApi.GetAllPersonsByGroupId(1, 10, _groupId.Value);
            var initialPersonIds = initialPersonsResponse.Items?.Select(person => person.Id).ToList();

            // Removing them by updating group
            var updateGroup = new UpdateGroup
            {
                RemoveItems = initialPersonIds
            };
            PathsConfig.FaceSdk.GroupApi.UpdatePersonsInGroup(_groupId.Value, updateGroup);

            // Creating new persons and adding them to the group
            var newPersonNames = new List<string> { "New Person A", "New Person B" };
            var newPersonIds = new List<Guid>();
            var index = 0;
            for (; index < newPersonNames.Count; index++)
            {
                var name = newPersonNames[index];
                var newPersonFields = new PersonFields
                {
                    Name = name,
                    Groups = new List<Guid> { _groupId.Value },
                    Metadata = _baseMetadata
                };
                var newPersonId = PathsConfig.FaceSdk.PersonApi.CreatePerson(newPersonFields).Id;
                newPersonIds.Add(newPersonId);
            }

            // Testing that they're here
            var updatedPersonsResponse = PathsConfig.FaceSdk.GroupApi.GetAllPersonsByGroupId(1, 10, _groupId.Value);
            var updatedPersonIds = updatedPersonsResponse.Items?.Select(person => person.Id).ToList();
            var updatedPersonNames = updatedPersonsResponse.Items?.Select(person => person.Name).ToList();

            // Sort the arrays before comparing
            newPersonIds.Sort();
            updatedPersonIds?.Sort();
            newPersonNames.Sort();
            updatedPersonNames?.Sort();

            Assert.That(updatedPersonIds, Is.EqualTo(newPersonIds), "Updated person ids don't match");
            Assert.That(updatedPersonNames, Is.EqualTo(newPersonNames), "Updated person names don't match");
            Assert.That(updatedPersonIds?.Count, Is.EqualTo(2),
                "Expected 2 persons in updated list but got a different number");
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.Message);
        }
    }


    [Test]
    public void ShouldCreateGroupWithoutMetadata()
    {
        var groupToCreate = new GroupToCreate
        {
            Name = TestGroupName
        };
        _groupResponse = PathsConfig.FaceSdk.GroupApi.CreateGroup(groupToCreate);
        var newGroupId = _groupResponse.Id;
        PathsConfig.FaceSdk.GroupApi.DeleteGroup(newGroupId);
    }
}